sys/usb/usbus/cdc/acm/cdc_acm_uart.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usb_acm Virtual Serial Port UART implementation
+ * @{
+ *
+ * @file
+ * @brief CDC ACM UART implementation
+ *
+ * This file implements a USB CDC ACM callback and read functions from an UART
+ * peripheral.
+ * @}
+ */
+
+#include "periph/uart.h"
+#include "usb/cdc.h"
+#include "usb/usbus/cdc/acm_uart.h"
+
+#ifdef PERIPH_UART_MODECFG
+static int _acm2uart_bits(uint8_t bits, uart_data_bits_t *conv)
+{
+    switch (bits) {
+        case 5:
+            *conv = UART_DATA_BITS_5;
+            break;
+        case 6:
+            *conv = UART_DATA_BITS_6;
+            break;
+        case 7:
+            *conv = UART_DATA_BITS_7;
+            break;
+        case 8:
+            *conv = UART_DATA_BITS_8;
+            break;
+        default:
+            return -1;
+    }
+    return 0;
+}
+
+static int _acm2uart_parity(uint8_t parity, uart_parity_t *conv)
+{
+    switch (parity) {
+        case USB_CDC_ACM_CODING_PARITY_NONE:
+            *conv = UART_PARITY_NONE;
+            break;
+        case USB_CDC_ACM_CODING_PARITY_EVEN:
+            *conv = UART_PARITY_EVEN;
+            break;
+        case USB_CDC_ACM_CODING_PARITY_ODD:
+            *conv = UART_PARITY_ODD;
+            break;
+        case USB_CDC_ACM_CODING_PARITY_MARK:
+            *conv = UART_PARITY_MARK;
+            break;
+        case USB_CDC_ACM_CODING_PARITY_SPACE:
+            *conv = UART_PARITY_SPACE;
+            break;
+        default:
+            return -1;
+    }
+    return 0;
+}
+
+static int _acm2uart_stop(uint8_t stop, uart_stop_bits_t *conv)
+{
+    switch (stop) {
+        case USB_CDC_ACM_CODING_STOP_BITS_1:
+            *conv = UART_STOP_BITS_1;
+            break;
+        case USB_CDC_ACM_CODING_STOP_BITS_2:
+            *conv = UART_STOP_BITS_2;
+            break;
+        default:
+            return -1;
+    }
+    return 0;
+}
+#endif
+
+static void _rx_cb(void *arg, uint8_t data)
+{
+    usbus_cdc_acm_uart_device_t *acmuart = (usbus_cdc_acm_uart_device_t *)arg;
+
+    usbus_cdc_acm_submit(&acmuart->cdcacm, &data, 1);
+    if (data == '\n') {
+        usbus_cdc_acm_flush(&acmuart->cdcacm);
+    }
```

Comment:
How about transmit byte(s) as soon as they arrive in order to be more like a USB-to-UART adapter ?

Reply:
Those send chunks of bytes too

Reply:
@koen is there any timeout here, or would this essentially be line based?

Reply:
Something timeout-based would be better here indeed. I'll try to come up with something.

Reply:
I would imagine you could receive bytes on the UART faster than you can submit USB packets. 

So for the best performance, you would have double-buffering and 

 - fill the RX buffer in the UART callback
 - in a thread, check if data is in the RX buffer, if so swap buffers and transfer the buffer over USB

Reply:
So it seems line based. Is it very inefficient to flush every byte? It probably is, so some timeout would be needed.

Somethink like:
1. flush on every LF ('\r' instead of '\n', as terminals either send LF or CRLF).
2. flush every N bytes (16, 32, ...)
3. flush every M ms (with M being about the time it takes to transfer N bytes?)

maybe 3. is sufficient?

Chunking *might* cause problems, if the sender sends with a deliberate pause to not overload a shitty UART implementation, but the adapter sends full-speed N byte bursts...

Reply:
I think I prefer the flush after timeout and possible after every newline.